python/simulation.py

```diff
@@ -2284,10 +2312,16 @@ def add_source(self, src):
             ]
             add_eig_src = functools.partial(self.fields.add_eigenmode_source, *add_eig_src_args)
 
-            if src.amp_func is None:
-                add_eig_src()
+            if isinstance(src.eig_band, DiffractedPlanewave):
+                if src.amp_func is None:
+                    add_eig_src(lambda v: 1.0, diffractedplanewave)
+                else:
+                    add_eig_src(src.amp_func, diffractedplanewave)
```

Comment:
You should just be able to pass `src.amp_func` even if it is `None`, if you [edit line 699 of meep.i](https://github.com/NanoComp/meep/blob/98a49d09863c57062df2f239123bb0b4587decc1/python/meep.i#L699 to
```
$1 = PyCallable_Check($input) || $input == Py_None;
```
(then you don't even need the `if src.amp_func is None`)

Reply:
You'll also need to change [this typemap](https://github.com/NanoComp/meep/blob/98a49d09863c57062df2f239123bb0b4587decc1/python/meep.i#L703) to:
```
%typemap(in) std::complex<double> (*)(const meep::vec &) {
    if ($input) {
        $1 = py_amp_func_wrap;
        py_amp_func = $input;
        Py_INCREF(py_amp_func);
    }
    else
        $1 = NULL;
}
```